swarm/src/executor.rs

```diff
@@ -0,0 +1,59 @@
+use futures::executor::ThreadPool;
+use std::{future::Future, pin::Pin};
+
+/// Implemented on objects that can run a `Future` in the background.
+///
+/// > **Note**: While it may be tempting to implement this trait on types such as
+/// >           [`futures::stream::FuturesUnordered`], please note that passing an `Executor` is
+/// >           optional, and that `FuturesUnordered` (or a similar struct) will automatically
+/// >           be used as fallback by libp2p. The `Executor` trait should therefore only be
+/// >           about running `Future`s in the background.
+pub trait Executor {
+    /// Run the given future in the background until it ends.
+    fn exec(&self, future: Pin<Box<dyn Future<Output = ()> + Send>>);
+}
+
+impl<F: Fn(Pin<Box<dyn Future<Output = ()> + Send>>)> Executor for F {
+    fn exec(&self, f: Pin<Box<dyn Future<Output = ()> + Send>>) {
+        self(f)
+    }
+}
+
+impl Executor for ThreadPool {
+    fn exec(&self, future: Pin<Box<dyn Future<Output = ()> + Send>>) {
+        self.spawn_ok(future)
+    }
+}
+
+#[cfg(feature = "tokio")]
+#[derive(Default, Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) struct TokioExecutor;
+
+#[cfg(feature = "tokio")]
+impl Executor for TokioExecutor {
+    fn exec(&self, future: Pin<Box<dyn Future<Output = ()> + Send>>) {
+        let _ = tokio::spawn(future);
+    }
+}
+
+#[cfg(feature = "async-std")]
+#[derive(Default, Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) struct AsyncStdExecutor;
+
+#[cfg(feature = "async-std")]
+impl Executor for AsyncStdExecutor {
+    fn exec(&self, future: Pin<Box<dyn Future<Output = ()> + Send>>) {
+        let _ = async_std::task::spawn(future);
+    }
+}
+
+#[cfg(feature = "wasm-bindgen")]
+#[derive(Default, Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) struct WasmBindgenExecutor;
+
+#[cfg(feature = "wasm-bindgen")]
+impl Executor for WasmBindgenExecutor {
+    fn exec(&self, future: Pin<Box<dyn Future<Output = ()> + Send>>) {
+        wasm_bindgen_futures::spawn_local(future)
```

Comment:
Does this end up being any different from using our local queue?

It would be cool if this could use the web-worker API for example but I am not sure how easy this is.

Reply:
>  Does this end up being any different from using our local queue?

Reply:
> > Does this end up being any different from using our local queue?
> 
> Good question. I think yes, although I'm not an expert here. As far as I understand it, spawn_local submits a task to the browser's native background task queue. Which might enable the browser to schedule more cleverly than we can do. We would probably have to try it or have an expert comment. Maybe @melekes?
>
> 
> > It would be cool if this could use the web-worker API for example but I am not sure how easy this is.
> 
> Yes, it would be, but it's really hard as I see it. There are some attempts to build something like this for Rust and although possible the main issue stays: When starting a Web Worker one has to give a path to a JS file which then creates the worker. Secondly sending to and from Web Workers is really difficult. I think it requires serialization. Maybe I'm wrong here. I would be happy to be corrected.
> 
> Of course, it would be better if WASM could use threads directly. That used to be possible but got really difficult because of the changes made to [SharedArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer#security_requirements). Also: You can bring the Rust compiler to allow for multithreading but you have to compile the standard library yourself: [Parallel Raytracing Example](https://rustwasm.github.io/docs/wasm-bindgen/examples/raytrace.html).
> 
> Further reading: [tokio-rs/tokio#4827](https://github.com/tokio-rs/tokio/issues/4827)

Thanks for linking all this. I am happy to merge this executor.